File: scripts/update_stats.py
# -*- coding: utf-8 -*-
import os, sys, json, re, urllib.request
from datetime import datetime

def fetch_json(url_str, headers=None):
    if headers is None:
        headers = {}
    headers.setdefault('User-Agent', 'Python-Stats-Sync')
    req = urllib.request.Request(url_str, headers=headers)
    try:
        with urllib.request.urlopen(req, timeout=10) as resp:
            if resp.status == 200:
                return json.loads(resp.read().decode('utf-8'))
    except Exception as e:
        print(f"Aviso al consultar {url_str}: {e}")
    return None

def get_stats():
    current_year = datetime.now().year
    start_year = 2017
    years_exp = max(1, current_year - start_year)
    age_text = f"+ {years_exp} Años Como Solo Developer"

    # 1. Obtener juegos de EternoDev
    games_count = 5
    games_data = fetch_json("https://raw.githubusercontent.com/paucg06/EternoDev/main/games.json")
    if isinstance(games_data, list) and len(games_data) > 0:
        games_count = len(games_data)

    # 2. Obtener apps escaneando temas/topics de GitHub o fallback
    apps_count = 5
    gh_token = os.environ.get("GITHUB_TOKEN")
    gh_headers = {"Authorization": f"Bearer {gh_token}"} if gh_token else {}

    try:
        user_repos = fetch_json("https://api.github.com/users/paucg06/repos?per_page=100", gh_headers)
        tagged_apps = 0
        if isinstance(user_repos, list):
            app_topics = {"app", "mobile-app", "android-app", "ios-app", "application", "mobile"}
            for repo in user_repos:
                if not repo.get("fork", False):
                    topics = set(t.lower() for t in repo.get("topics", []))
                    if topics & app_topics:
                        tagged_apps += 1
        if tagged_apps > 0:
            apps_count = tagged_apps
    except Exception as e:
        print(f"Aviso al consultar repositorios de GitHub: {e}")

    return {
        "years_exp": years_exp,
        "age_text": age_text,
        "games_count": f"{games_count:02d}",
        "apps_count": f"{apps_count:02d}"
    }

def update_svg_file(file_path, stats):
    if not os.path.exists(file_path):
        return
    with open(file_path, "r", encoding="utf-8") as f:
        content = f.read()

    # Total ancho línea = 59 caracteres monospaced
    # 1. Actividad (12 chars prefix): '. Actividad:'
    age_prefix_len = 12
    age_dots_count = max(2, 59 - age_prefix_len - 1 - len(stats["age_text"]))
    age_dots = " " + ("." * age_dots_count) + " "

    content = re.sub(
        r'<tspan class="cc" id="age_data_dots">[^<]*<\/tspan><tspan class="value" id="age_data">[^<]*<\/tspan>',
        f'<tspan class="cc" id="age_data_dots">{age_dots}</tspan><tspan class="value" id="age_data">{stats["age_text"]}</tspan>',
        content
    )

    # 2. Apps Creadas (15 chars prefix): '. Apps Creadas:'
    apps_prefix_len = 15
    apps_dots_count = max(2, 59 - apps_prefix_len - 1 - len(stats["apps_count"]))
    apps_dots = " " + ("." * apps_dots_count) + " "

    content = re.sub(
        r'<tspan class="cc" id="repo_data_dots">[^<]*<\/tspan><tspan class="value" id="repo_data">[^<]*<\/tspan>',
        f'<tspan class="cc" id="repo_data_dots">{apps_dots}</tspan><tspan class="value" id="repo_data">{stats["apps_count"]}</tspan>',
        content
    )

    # 3. Juegos Creados (17 chars prefix): '. Juegos Creados:'
    games_prefix_len = 17
    games_dots_count = max(2, 59 - games_prefix_len - 1 - len(stats["games_count"]))
    games_dots = " " + ("." * games_dots_count) + " "

    content = re.sub(
        r'<tspan class="cc" id="commit_data_dots">[^<]*<\/tspan><tspan class="value" id="commit_data">[^<]*<\/tspan>',
        f'<tspan class="cc" id="commit_data_dots">{games_dots}</tspan><tspan class="value" id="commit_data">{stats["games_count"]}</tspan>',
        content
    )

    with open(file_path, "w", encoding="utf-8") as f:
        f.write(content)
    print(f"Actualizado: {file_path}")

def main():
    print("Obteniendo estadísticas dinámicas...")
    stats = get_stats()
    print("Estadísticas calculadas:", stats)

    base_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    update_svg_file(os.path.join(base_dir, "dark_mode.svg"), stats)
    update_svg_file(os.path.join(base_dir, "light_mode.svg"), stats)
    print("¡Sincronización completada con éxito!")

if __name__ == "__main__":
    main()

File: scripts/update_stats.js
const fs = require('fs');
const https = require('https');
const path = require('path');

function fetchJson(urlStr, headers = {}) {
  return new Promise((resolve) => {
    const options = {
      headers: {
        'User-Agent': 'NodeJS-Stats-Sync',
        ...headers
      }
    };
    https.get(urlStr, options, (res) => {
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        return resolve(fetchJson(res.headers.location, headers));
      }
      if (res.statusCode !== 200) {
        return resolve(null);
      }
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        try {
          resolve(JSON.parse(data));
        } catch (e) {
          resolve(null);
        }
      });
    }).on('error', () => resolve(null));
  });
}

async function getStats() {
  const currentYear = new Date().getFullYear();
  const startYear = 2017;
  const yearsExp = Math.max(1, currentYear - startYear);
  const ageText = `+ ${yearsExp} Años Como Solo Developer`;

  // 1. Obtener juegos de EternoDev
  let gamesCount = 7; // Valor de respaldo por defecto
  const gamesData = await fetchJson('https://raw.githubusercontent.com/paucg06/EternoDev/main/games.json');
  if (Array.isArray(gamesData) && gamesData.length > 0) {
    gamesCount = gamesData.length;
  }

  // 2. Obtener apps escaneando temas/topics de GitHub o fallback
  let appsCount = 5;
  const githubHeaders = process.env.GITHUB_TOKEN ? { 'Authorization': `Bearer ${process.env.GITHUB_TOKEN}` } : {};
  
  try {
    const userRepos = await fetchJson('https://api.github.com/users/paucg06/repos?per_page=100', githubHeaders);
    let taggedApps = 0;
    if (Array.isArray(userRepos)) {
      const appTopics = ['app', 'mobile-app', 'android-app', 'ios-app', 'application', 'mobile'];
      for (const repo of userRepos) {
        if (!repo.fork) {
          const topics = repo.topics || [];
          if (topics.some(t => appTopics.includes(t.toLowerCase()))) {
            taggedApps++;
          }
        }
      }
    }
    if (taggedApps > 0) {
      appsCount = taggedApps;
    }
  } catch (e) {
    console.warn('No se pudieron consultar repos de GitHub:', e.message);
  }

  return {
    yearsExp,
    ageText,
    gamesCount: String(gamesCount).padStart(2, '0'),
    appsCount: String(appsCount).padStart(2, '0')
  };
}

function updateSvgFile(filePath, stats) {
  if (!fs.existsSync(filePath)) return;
  let content = fs.readFileSync(filePath, 'utf-8');

  // Total de ancho de línea = 59 caracteres
  // 1. Actividad: '. Actividad:' (12 chars) + dots + ' ' + ageText
  const agePrefixLen = 12; // ". Actividad:"
  const ageDotsCount = Math.max(2, 59 - agePrefixLen - 1 - stats.ageText.length);
  const ageDots = ' ' + '.'.repeat(ageDotsCount) + ' ';

  content = content.replace(
    /<tspan class="cc" id="age_data_dots">[^<]*<\/tspan><tspan class="value" id="age_data">[^<]*<\/tspan>/,
    `<tspan class="cc" id="age_data_dots">${ageDots}</tspan><tspan class="value" id="age_data">${stats.ageText}</tspan>`
  );

  // 2. Apps Creadas: '. Apps Creadas:' (15 chars) + dots + ' ' + appsCount
  const appsPrefixLen = 15;
  const appsDotsCount = Math.max(2, 59 - appsPrefixLen - 1 - stats.appsCount.length);
  const appsDots = ' ' + '.'.repeat(appsDotsCount) + ' ';

  content = content.replace(
    /<tspan class="cc" id="repo_data_dots">[^<]*<\/tspan><tspan class="value" id="repo_data">[^<]*<\/tspan>/,
    `<tspan class="cc" id="repo_data_dots">${appsDots}</tspan><tspan class="value" id="repo_data">${stats.appsCount}</tspan>`
  );

  // 3. Juegos Creados: '. Juegos Creados:' (17 chars) + dots + ' ' + gamesCount
  const gamesPrefixLen = 17;
  const gamesDotsCount = Math.max(2, 59 - gamesPrefixLen - 1 - stats.gamesCount.length);
  const gamesDots = ' ' + '.'.repeat(gamesDotsCount) + ' ';

  content = content.replace(
    /<tspan class="cc" id="commit_data_dots">[^<]*<\/tspan><tspan class="value" id="commit_data">[^<]*<\/tspan>/,
    `<tspan class="cc" id="commit_data_dots">${gamesDots}</tspan><tspan class="value" id="commit_data">${stats.gamesCount}</tspan>`
  );

  fs.writeFileSync(filePath, content, 'utf-8');
  console.log(`Actualizado ${filePath}`);
}

async function main() {
  console.log('Obteniendo estadísticas...');
  const stats = await getStats();
  console.log('Estadísticas calculadas:', stats);

  const basePath = path.join(__dirname, '..');
  updateSvgFile(path.join(basePath, 'dark_mode.svg'), stats);
  updateSvgFile(path.join(basePath, 'light_mode.svg'), stats);
  console.log('¡Sincronización completada!');
}

main();
